src/Rcl/SillyArrayMap.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE TypeApplications #-}
+
+-- | Stackage changed its JSON representation for package differences
+--
+-- It used to be nested objects:
+--
+-- @
+-- { diff:
+--   { some-package:    { lts-x: version-x, lts-y: version-y }
+--   , another-package: { lts-x: version-x, lts-y: version-y }
+--   , ...
+--   }
+-- }
+-- @
+--
+-- which was very easy to work with in Haskell (map of maps).
+--
+-- Now it's this:
+--
+-- @
+-- { diff:
+--   [ ["some-package",    { lts-x: version-x, lts-y: version-y }]
+--   , ["another-package", { lts-x: version-x, lts-y: version-y }]
+--   , ...
+--   ]
+-- }
+-- @
+--
+-- Which is /not/ easy to work with in Haskell (list of heterogenious list).
+--
+-- I guess is nicer for JavaScript? Any, I'm representing this silly type and
+-- JSON parser here, to isolate the mess. It allows a drop-in change in my main
+-- type to fix the parsing.
+--
+module Rcl.SillyArrayMap
+  ( SillyArrayMap
+  , toList
+  )
+where
+
+import RIO hiding (toList)
+
+import Data.Aeson
+import Data.Aeson.Types (Parser)
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.Vector as V
+import Unsafe.Coerce (unsafeCoerce)
+
+newtype SillyArrayMap k v = SillyArrayMap
+  { unSillyArrayMap :: HashMap k v
+  }
+
+toList :: SillyArrayMap k v -> [(k, v)]
+toList = HashMap.toList . unSillyArrayMap
+
+instance (Eq k, Hashable k, FromJSONKey k, FromJSON v) => FromJSON (SillyArrayMap k v) where
+  parseJSON =
+    withArray "SillyArrayMap"
+      $ fmap (SillyArrayMap . HashMap.fromList)
+      . traverse parseSillyArrayElem
+      . V.toList
+
+parseSillyArrayElem :: (FromJSONKey k, FromJSON v) => Value -> Parser (k, v)
+parseSillyArrayElem = withArray "SillyArrayMapElement" $ go . V.toList
+ where
+  go = \case
+    [k, v] -> (,) <$> parseJSONKey k <*> parseJSON v
+    _ -> fail "Silly Array element wasn't exactly two keys"
+
+parseJSONKey :: forall k . FromJSONKey k => Value -> Parser k
+parseJSONKey v = case fromJSONKey @k of
+  FromJSONKeyCoerce _ -> withText' $ pure . unsafeCoerce
```

Comment:
You should be able to use `coerce` here. The argument to `FromJSONKeyCoerce` carries around a `Coercible` dictionary at runtime. Not sure if you need to pattern match on it or if pattern matching on `FromJSONKeyCoerce` is enough evidence:

```haskell
data FromJSONKeyFunction a
    = FromJSONKeyCoerce !(CoerceText a)
      -- ^ uses 'coerce' ('unsafeCoerce' in older GHCs)
    | ...

data CoerceText a where
    CoerceText :: Coercible Text a => CoerceText a
```

Reply:
`aeson` [`master`](https://github.com/bos/aeson/blob/eef26cfad2d61e2425e3fc30c5adb07a9dc67d5f/Data/Aeson/Types/FromJSON.hs#L429) uses a `GADT` for `FromJSONKeyFunction` to avoid the auxiliary data type, kinda cool:

```haskell
data FromJSONKeyFunction a where
    FromJSONKeyCoerce :: Coercible Text a => FromJSONKeyFunction a
    FromJSONKeyText :: !(Text -> a) -> FromJSONKeyFunction a
    FromJSONKeyTextParser :: !(Text -> Parser a) -> FromJSONKeyFunction a
    FromJSONKeyValue :: !(Value -> Parser a) -> FromJSONKeyFunction a
```

Reply:
I wanted to use `coerce` but couldn't figure out how to make use of that dictionary. Do you have any more specifics (aka working code) that could guide me?

Reply:
> if pattern matching on FromJSONKeyCoerce is enough evidence

It was not, that was as far as I got and it didn't work.

Reply:
I think I can't do this because `CoerceText` is not exported.